warehouse/query-core/src/main/java/datawave/query/jexl/visitors/DelayedNonEventSubTreeVisitor.java

```diff
@@ -0,0 +1,132 @@
+            // IteratorBuildingVisitor
+            JexlNode candidate = QueryPropertyMarker.getQueryPropertySource(node, ASTDelayedPredicate.class);
+            
+            if (candidate == null) {
+                // we are in a recursive situation inside an delayed predicate, evaluate the node
+                candidate = node;
+            }
+            
+            extractDelayedFields(candidate);
+            
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            return super.visit(candidate, true);
+        }
+        
+        // it wasn't a delayed node, process down the tree
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        if (data != null && (boolean) data) {
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            extractDelayedFields(node);
+        }
+        
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        if (data != null && (boolean) data) {
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            extractDelayedFields(node);
+        }
+        
+        return super.visit(node, data);
+    }
+    
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        if (data != null && (boolean) data) {
+            // continue processing the tree in case the IteratorBuildingVisitor decided to discard leafs. This is not the most efficient or ideal
+            // but it avoids logic problems in AndIterator/OrIterator that haven't yet been fixed.
+            extractDelayedFields(node);
+        }
+        
+        return super.visit(node, data);
+    }
+    
+    private void extractDelayedFields(JexlNode node) {
+        // reset the root before processing a a node
+        iteratorBuildingVisitor.resetRoot();
+        
+        // use the iterator building visitor to build a field to node map for each delayed node
+        node.jjtAccept(iteratorBuildingVisitor, null);
+        NestedIterator<Key> root = iteratorBuildingVisitor.root();
+        if (root != null) {
+            for (NestedIterator<Key> leaf : root.leaves()) {
+                // only IndexIteratorBridge nodes matter, everything else can be ignored
```

Comment:
why?

Reply:
Any leaves that come out of this are delayed. Of those delayed leaves the only ones that require delayed fetching are those that are non-event fields. Everything else will be filled in from the event fetch.